### ISSUE TITLE
PyImport_Import crash

### ISSUE THREAD

**Issue by kanoncom** [tensorflow/tensorflow#11546]
I recently ran into a problem, while using bazel to build a project.  This project is compiled as a dynamic linking library, using PyImport_Import to import python module. when there is "import tensorflow as tf" in the python file , application who calls the dynamic linking library crashed everytime, but when it‘s not there ,everything works just fine. where is the problem?
my tensorflow version is 1.0.0,python 2.7.0,bazel 0.4.3

here is the console information when the application crashes:

> F tensorflow/core/framework/function.cc:1015] Check failed: GetOpGradFactory()->insert({op, func}).second Duplicated gradient for Softmax

here is the test python file looks like: 
`from __future__ import print_function
import tensorflow as tf
import os
import time
from itertools import izip
import numpy as np
import wrapt
import cv2
def get_int( ):
    a = 10
    b = 20
    return a + b

def get_str( s1, s2 ):
    #return s1 + s2  
    #return 'Hello , TY'  
    return ('Hello, World', 10, 20)`

here is the source code of .so file:
`   Py_Initialize();
    if ( !Py_IsInitialized() ) {
        return -1;
    }
    PyEval_InitThreads();
    PyThreadState *mainThreadState = NULL;
    //  save a pointer to the main PyThreadState object 
    mainThreadState = PyThreadState_Get();
    //  release the lock 
    PyEval_ReleaseLock();
    char* mockargv[1]={(char*)""};
    PySys_SetArgv(1,mockargv);
    PyRun_SimpleString("import sys");
    PyRun_SimpleString("sys.path.append('./')");
    pName_ = PyString_FromString("test_py");
    displayPyObject(pName_);
    if(pName_ == NULL){
        return -1;
    }
    pModule_ = PyImport_Import(pName_);
    displayPyObject(pModule_);`

and here is the dynamic linking library part of my BUILD file
`cc_binary(
    name = "test.so",
    linkshared = 1,
    deps = [
        ":test_lib",
    ],
)
cc_library(
    name = "test_lib",
    visibility = ["//visibility:__subpackages__"],
    srcs = glob(["test.cpp"
             ],
        ),
    includes=["test.h"],
    linkopts = [
"-lm -lpthread -L/usr/lib/python2.7 -lpython2.7 -lopencv_core -lopencv_imgproc -lopencv_highgui -lopencv_ml -lfreeimage"
    ],
    deps = [
        "//tensorflow/cc:cc_ops",
        "//tensorflow/core:framework",
        "//tensorflow/core:framework_internal",
        "//tensorflow/core:tensorflow",
    ],
)
`

**Comment by shreyneil** [comment]
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/docs_src/extend/adding_an_op.md
https://github.com/tensorflow/tensorflow/issues/7403
try these.

**Comment by asimshankar** [comment]
I apologize but I am having a hard time understanding what the problem is, where the problem is, and what version it affects. Please pay attention to the issue template (https://github.com/tensorflow/tensorflow/issues/new) and provide all the information it asks. Thank you.

That said, based on what you've suggested, I would suspect the segfault is because of dual initialization of some global state. As @shreyneil pointed out, if you're adding an op, please follow the documentation at https://www.tensorflow.org/extend/, which will result in your custom library extension linking against `pywrap_tensorflow.so` and avoiding any global state conflicts.

If that doesn't seem to be the case, please feel free to file a new issue with all requested details and most importantly detailed instructions to reproduce the problem. Hope that helps, thanks!
